sympy/solvers/inequalities.py

```diff
@@ -456,6 +456,27 @@ def solve_univariate_inequality(expr, gen, relational=True, domain=S.Reals, cont
     from sympy.solvers.solvers import denoms
     from sympy.solvers.solveset import solvify, solveset
 
+    if domain.is_subset(S.Reals) is False:
+        raise NotImplementedError(filldedent('''
+        Inequalities in the complex domain are
+        not supported. Try the real domain by
+        setting domain=S.Reals'''))
+    elif domain is not S.Reals:
+        rv = solve_univariate_inequality(
+        expr, gen, relational=False, continuous=continuous).intersection(domain)
+        if relational:
+            try:
+                rv = rv.as_relational(gen)
+            except NotImplementedError:
+                if isinstance(rv, Union):
+                    rv = Or(*[i.as_relational(gen) for i in rv.args])
+                else:
+                    raise NotImplementedError(filldedent('''
+                            Try changing the flag as 'relational=False'. '''))
```

Comment:
This looks interesting, @oscarbenjamin . Can you show me how you get it to work? I am not having success getting it to work. And under what circumstances do you envision it being slower?

Reply:
What is slow is something like this:
```python
from sympy import *

x = Symbol('x')

expr = 1 + x

for n in range(100):
    expr = expr * (1 + expr)

def f(expr):
    if not isinstance(expr, Add):
        raise NotImplementedError("""
            Expected an Add but found %s
                """ % expr)
    return 2

try:
    z = f(expr)
except NotImplementedError:
    z = 0

print(z)
```
It is slow because `% expr` calls `str(expr)` which is really slow for this large expression. That's completely wasted though because when we catch the exception we throw it (and its message away).

We can make it faster in the cases where we don't want to see the exception by delaying the `% expr` part:
```python
from sympy import *


class SymPyError(Exception):
    def __init__(self, message, *fmt):
        self.message = message
        self.fmt = fmt

    def __str__(self):
        return '%s(%s)' % (type(self).__name__, self.message % self.fmt)

class SymPyNotImplementedError(SymPyError, NotImplementedError):
    pass


x = Symbol('x')

expr = 1 + x

for n in range(100):
    expr = expr * (1 + expr)

def f(expr):
    if not isinstance(expr, Add):
        raise SymPyNotImplementedError("""
            Expected an Add but found %s
                """, expr)
    return 2

try:
    z = f(expr)
except NotImplementedError:
    z = 0

print(z)
```
We can also use that to call `filldedent` but only when the message is actually needed.